### PR TITLE
[ios] Inconsistent layout in place page on iOS

### PR DESCRIPTION
**Pull request by kirylkaveryn**
Closes https://github.com/organicmaps/organicmaps/issues/5088
Closes #3454

Problems:
1. **Unexpected grey spacing**
Bug caused by changing spacings in UISackView header in PlacePageViewController
```swift
  func addHeader(_ headerViewController: UIViewController) {
    addToStack(headerViewController)
    // TODO: workaround. Custom spacing isn't working if visibility of any arranged subview
    // changes after setting custom spacing
    DispatchQueue.main.async {
      self.stackView.setCustomSpacing(0, after: headerViewController.view)
    }
  }
```
Solving:
Change layout of header. Now place page header is array of VCs (see IPlacePageLayout protocol) in stack view with zero spacing and proper layout. And every PlacePageLayout object should build content of a header.
No runtime changing of spacings.

2. **unexpected "double line"** 
Separators in PlacePageInfo VC's stack view are adding on every view in storyboard manually so there is no option to disable current separator and this cause doubling in some places. 
Solving:
Remove separators from VC's views is SB and add them programmatically only when view is added to stack view. So UIStackView now behave like UITableView in case of separators making:
```swift
  private func addToStack(_ viewController: UIViewController) {
    addChild(viewController)
    stackView.addArrangedSubview(viewController.view,
                                 withSeparator: .default.withInsets(UIEdgeInsets(top: 0, left: 56, bottom: 0, right: 0)))
    viewController.didMove(toParent: self)
  }
```
IMG_2260

3. **overly large space after Wikipedia article and no lines at top and bottom of article**
Solving: fix contentHuggingPriority for wiki's UITextView and add top/bottom separators
image
4. New bug was found and fixed: when user tap on "more" in wiki info VC to see additional info, and than goes back, style of wiki info became broken (see gif).
Simulator Screen Recording - iPhone 14 Pro Max - 2023-10-04 at 13 42 58

This happens because there was some User Defined Runtime Attributes set in storyboard and they override styles at runtime.
image

### CODE REVIEW

**Comment by dvdmrtnz** [comment]
Looks good! The grey spacing that showed up randomly was quite annoying! Nice job!

File: iphone/Maps/UI/PlacePage/Components/PlacePageInfoViewController.swift [hunk 1]
```diff
@@ -239,7 +236,8 @@ class PlacePageInfoViewController: UIViewController {
 
   private func addToStack(_ viewController: UIViewController) {
     addChild(viewController)
-    stackView.addArrangedSubview(viewController.view)
+    stackView.addArrangedSubview(viewController.view,
+                                 withSeparator: .default.withInsets(UIEdgeInsets(top: 0, left: 56, bottom: 0, right: 0)))
```
Comment (biodranik) [inline iphone/Maps/UI/PlacePage/Components/PlacePageInfoViewController.swift:240]: How 56 was chosen?

File: iphone/Maps/Categories/UIStackView+ArrangedSubviewWithSeparator.swift [hunk 2]
```diff
@@ -0,0 +1,103 @@
+extension UIStackView {
```
Comment (biodranik) [inline iphone/Maps/Categories/UIStackView+ArrangedSubviewWithSeparator.swift:1]: I'm curious, is it the simplest method to fix the issue?

File: iphone/Maps/Categories/UIStackView+ArrangedSubviewWithSeparator.swift [hunk 3]
```diff
@@ -40,64 +44,20 @@ extension UIStackView {
...
-    case leading, trailing, top, bottom
+  /// A default value returns separator configuration with a thickness of 1.0, zero insets, position "before" and color fetched from current system style.
+  static var `default`: Self {
+    .init(thickness: CGFloat(1.0),
+          color: StyleManager.shared.theme?.colors.blackDividers,
+          insets: .zero,
+          position: .before)
   }
   
-  func addLine(thickness: CGFloat,
```
Comment (biodranik) [inline iphone/Maps/Categories/UIStackView+ArrangedSubviewWithSeparator.swift]: Looks like this commit can be squashed with the first one, right?

Do you have experience with git rebase/fixup/squashing and partial commit editing to clean up commit history?

Btw, [Fork](https://git-fork.com/) app is the best Git client for Mac.

File: iphone/Maps/Maps.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -2030,6 +2034,8 @@
 		3454D7981E07F045004AF2AD /* Categories */ = {
 			isa = PBXGroup;
 			children = (
+				ED8AC9092ACD9DA100EECC90 /* UIStackView+ArrangedSubviewWithSeparator.swift */,
+				ED8AC90C2ACD9DA100EECC90 /* UIView+AddLine.swift */,
```
Comment (biodranik) [inline iphone/Maps/Maps.xcodeproj/project.pbxproj:2038]: Should these additions be a part of another commit?

File: iphone/Maps/UI/Settings/Cells/SettingsTableViewLinkCell.swift [hunk 5]
```diff
@@ -2,18 +2,25 @@
 final class SettingsTableViewLinkCell: MWMTableViewCell {
   @IBOutlet fileprivate weak var title: UILabel!
   @IBOutlet fileprivate weak var info: UILabel?
+  
+  private let chevronImageView: UIImageView = {
+    let imageView = UIImageView(image: UIImage(named: "ic_arrow_gray_right"))
+    imageView.size = CGSize(width: 11, height: 28)
```
Comment (biodranik) [inline iphone/Maps/UI/Settings/Cells/SettingsTableViewLinkCell.swift:8]: How these sizes were calculated?

File: iphone/Maps/UI/Settings/MWMSettingsViewController.mm [hunk 6]
```diff
@@ -29,7 +29,7 @@ @interface MWMSettingsViewController ()<SettingsTableViewSwitchCellDelegate>
 @property(weak, nonatomic) IBOutlet SettingsTableViewSwitchCell *autoZoomCell;
 @property(weak, nonatomic) IBOutlet SettingsTableViewLinkCell *voiceInstructionsCell;
 @property(weak, nonatomic) IBOutlet SettingsTableViewLinkCell *drivingOptionsCell;
-
+@property(copy, nonatomic) NSString *headerIdentifier;
```
Comment (biodranik) [inline iphone/Maps/UI/Settings/MWMSettingsViewController.mm:32]: Where is it used?

**Review by biodranik — COMMENTED** [review]
Thank you very much for your work! It looks great!

nit: It's better to avoid splitting commit titles into several lines. Necessary details can go into a commit's body.

Did you test it in portrait/landscape, iPad, and dark/light mode?

File: iphone/Maps/UI/PlacePage/Components/PlacePageInfoViewController.swift (hunk 1 above)
Comment (kirylkaveryn) [reply iphone/Maps/UI/PlacePage/Components/PlacePageInfoViewController.swift:240]: 56 is used, because every UIImageView in PlacePageInfo use this width and separators was constrained to them.

SCR-20231005-lvdu

File: iphone/Maps/Categories/UIStackView+ArrangedSubviewWithSeparator.swift (hunk 2 above)
Comment (kirylkaveryn) [reply iphone/Maps/Categories/UIStackView+ArrangedSubviewWithSeparator.swift:1]: The problem was not easy to fix because each view controller in PlacePageInfo contains own separator on bottom with constraints (it is not very comfortable to work with):
SCR-20231005-lute

So there are some approaches to fix that that i found:
1. hardcode each view when it is adding to the stack and try to inspect is it have separator or not, cast it and change it's visibility (a lot of hardcoding)
2. to rewrite all screen with UITableView (very hard)
3. to move responsibility of adding separators from each cells to stack view, remove separators from cells at all and use "tableview's style approach" for stack view then you just setup separator at one place
```swift
let tableView = UITableView()
tableView.separatorInset = UIEdgeInsets.zero
```
So i choose 3rd one because it is a bad idea to add separator to each view manually in storyboard and constraint them to some views. It is also not responsibility of cell to add and layout separators because cell shouldn't know its position in collection/stack. 
Now if we want to add some new info or rearrange PlacePage developer should not think about adding of separators manually and layouting them. And style of separator can be changed from **one place** in some clicks, but not in every view controller as before.
---
PS: I can move this code to private extension to PlacePage VC but i think that it will be Ok to use same approch if needed in some other places where we use UIStackView instead of UITableView.

File: iphone/Maps/Categories/UIStackView+ArrangedSubviewWithSeparator.swift (hunk 3 above)
Comment (kirylkaveryn) [reply iphone/Maps/Categories/UIStackView+ArrangedSubviewWithSeparator.swift]: Sorry for that, i'm using fork and cleaned up commit history from probably 12 to 5 but miss that... will fix that!

File: iphone/Maps/Maps.xcodeproj/project.pbxproj (hunk 4 above)
Comment (kirylkaveryn) [reply iphone/Maps/Maps.xcodeproj/project.pbxproj:2038]: My mistake. I will fix that!

File: iphone/Maps/UI/Settings/Cells/SettingsTableViewLinkCell.swift (hunk 5 above)
Comment (kirylkaveryn) [reply iphone/Maps/UI/Settings/Cells/SettingsTableViewLinkCell.swift:8]: The size of disclosure indicator that is used by default in table view's cells when selected accessory type is 11x14pt (i have checked in different ios version and screen sizes), but the problem with tis indicator that you jest cannon change its tint color (well know problem in ios), and you should use your own accory Image.
So i found that same already used in assets - ic_arrow_gray_right, but it needs to be resized to have same look as default. Not best approach but there is no other (except implementing of using system theme dark/light modes (from ios13) that is automatically handle image tint color)

File: iphone/Maps/UI/Settings/MWMSettingsViewController.mm (hunk 6 above)
Comment (kirylkaveryn) [reply iphone/Maps/UI/Settings/MWMSettingsViewController.mm:32]: My mistake :( . Will fix that
Thanks for being attentive!

**Comment by kirylkaveryn** [comment]
@biodranik 
I've tested it on iPhone 14pro (ios16.4), SE 2nd gen (ios14.1), iPad 11" ( ios16.4) and iPad for Mac
But it needs to be tested on ios12 and some other devices!

Here is the some gifs:
Simulator Screen Recording - iPhone SE (2nd generation) - 2023-10-05 at 14 45 16
Simulator Screen Recording - iPhone SE (2nd generation) - 2023-10-05 at 14 44 42
Simulator Screen Recording - iPhone 14 Pro - 2023-10-05 at 14 49 44
Simulator Screen Recording - iPad Pro (11-inch) (4th generation) - 2023-10-05 at 14 50 41

**Comment by kirylkaveryn** [comment]
@biodranik 
I've tested it on iPhone 14pro (ios16.4), SE 2nd gen (ios14.1), iPad 11" ( ios16.4) and iPad for Mac
But it needs to be tested on ios12 and some other devices!

Here is the some gifs:
Simulator Screen Recording - iPhone SE (2nd generation) - 2023-10-05 at 14 45 16
Simulator Screen Recording - iPhone SE (2nd generation) - 2023-10-05 at 14 44 42
Simulator Screen Recording - iPhone 14 Pro - 2023-10-05 at 14 49 44
Simulator Screen Recording - iPad Pro (11-inch) (4th generation) - 2023-10-05 at 14 50 41

File: iphone/Maps/Categories/UIStackView+ArrangedSubviewWithSeparator.swift (hunk 2 above)
Comment (biodranik) [reply iphone/Maps/Categories/UIStackView+ArrangedSubviewWithSeparator.swift:1]: Thanks for the explanation! Why is the table view approach more complex than the stack view one?

P.S. The KISS principle suggests avoiding doing unnecessary work if it's not needed ) If it's used now only in PP, let's move it there.

File: iphone/Maps/UI/Settings/Cells/SettingsTableViewLinkCell.swift (hunk 5 above)
Comment (biodranik) [reply iphone/Maps/UI/Settings/Cells/SettingsTableViewLinkCell.swift:8]: 1. Adding this comment into the code would be beneficial for future readers ;-)
2. Can we use [this SVG](https://fonts.google.com/icons?selected=Material%20Symbols%20Outlined%3Aarrow_forward_ios%3AFILL%400%3Bwght%40400%3BGRAD%400%3Bopsz%4024) or [any other SVG](https://fonts.google.com/icons?icon.query=right+&icon.platform=web) instead of resizing a pixel image?
3. Can we reuse iOS 13 theme features on iOS 13+ devices conditionally? Or would it be significantly simpler/easier to drop iOS 12?

File: iphone/Maps/Categories/UIStackView+ArrangedSubviewWithSeparator.swift (hunk 2 above)
Comment (kirylkaveryn) [reply iphone/Maps/Categories/UIStackView+ArrangedSubviewWithSeparator.swift:1]: TableView approach is easy to implement when building from scratch but in our case it needs to do some refactoring:
- changing layout and outlets in storyboard
- creating custom empty cells that will contain view controllers 
- registering/dequeuing of cells
- implement UITableViewDelegate and UITableViewDataSource protocols
- inserting resulting self-sizing non-scrollable UITableView inside of stackView

And we also needs to:
- remove separators from storyboard like i do
- set separators insets manually in uitableview

So it can be done but needs more time to refactor than just add separators manually

File: iphone/Maps/UI/Settings/Cells/SettingsTableViewLinkCell.swift (hunk 5 above)
Comment (kirylkaveryn) [reply iphone/Maps/UI/Settings/Cells/SettingsTableViewLinkCell.swift:8]: 1. thanks, got it!
2. yes! I tried not to add new assets, but if it ok - i'll add apples chevron default svg from apples SFSymblols
3. because of we are using custom Style system (with MWMThemeManager, MWMSettings, MWMRenderers and so on) when we tap on changing theme cell is Settings we are **actually do not change** UIKit's native (from ios13)`traitCollection.userInterfaceStyle` ... UIKit theme system work with same approach: goes through views hierarchy and change views colors. So both systems do the same work and mix them may cause unexpected behaviour... And i really don't find the way to use both systems for some views: whole design system looks very complex and hard to debug now...

Comment (biodranik) [reply iphone/Maps/UI/Settings/Cells/SettingsTableViewLinkCell.swift:8]: If the visual difference is negligible, the existing icon is also ok. But otherwise, you can import/use any better icons (just make sure that they work on iOS 12).

Regarding the dark theme, let's create another issue (or find an appropriate existing one) to discuss how to refactor it. For example, would it be very easy for us to support if we implement only the system theme support starting from iOS 13, and leave iOS 12 users without the dark theme? What are the other pros and cons of dropping iOS 12 completely?

**Comment by Jean-BaptisteC** [comment]
Can you add on your first comment `closes #3454`

**Comment by kirylkaveryn** [comment]
@biodranik I've made some updates in PR:
- removed fixing of Settings screen in the dark mode (will create it in separate PR)
- removed extensions to PP vc and made them private for local usage 
- squashed all commits into one

File: iphone/Maps/UI/PlacePage/PlacePage.storyboard [hunk 7]
```diff
@@ -510,7 +530,7 @@
                                 <rect key="frame" x="16" y="16" width="343" height="224"/>
                                 <subviews>
                                     <button opaque="NO" contentMode="scaleToFill" contentHorizontalAlignment="center" contentVerticalAlignment="center" buttonType="system" lineBreakMode="middleTruncation" translatesAutoresizingMaskIntoConstraints="NO" id="RUc-Qr-8Sb">
-                                        <rect key="frame" x="0.0" y="0.0" width="343" height="164"/>
+                                        <rect key="frame" x="0.0" y="0.0" width="343" height="44"/>
```
Comment (biodranik) [inline iphone/Maps/UI/PlacePage/PlacePage.storyboard:533]: nit: it would be great to skip non-relevant changes, or drop them completely.

**Review by biodranik — APPROVED** [review]

File: iphone/Maps/UI/PlacePage/PlacePage.storyboard (hunk 7 above)
Comment (kirylkaveryn) [reply iphone/Maps/UI/PlacePage/PlacePage.storyboard:533]: This change (and some others) happened because I resolved some constraint conflicts which was detected by xcode in storyboard (marked as red dots): 164 is incorrect
image

Comment (biodranik) [reply iphone/Maps/UI/PlacePage/PlacePage.storyboard:533]: Ok, hope it didn't break something )
